In [2]:
import matplotlib.pyplot as plt
from sklearn import linear_model
import statsmodels.formula.api as smf
import pandas as pd

# Display preferences.
%matplotlib inline
pd.options.display.float_format = '{:.3f}'.format

import warnings
warnings.filterwarnings(action="ignore")

In [4]:
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain            9.472   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain            9.356   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain            9.378   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain            8.289   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain            8.756   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                     7.389     0.890             14.120   
1                     7.228     0.860             14.265   
2                     9.378     0.890              3.928   
3                     5.944     0.830             14.104   
4                     6.978     0.830             11.045   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                 251.000           15.826       0.000              1015.130   
1                 259.000           15.826       0.000              1015.630   
2                 204.000           14.957       0.000              1015.940   
3                 269.000           15.826       0.000              1016.410   
4                 259.000           15.826       0.000              1016.510   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [5]:
weather_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Formatted Date            96453 non-null  object 
 1   Summary                   96453 non-null  object 
 2   Precip Type               95936 non-null  object 
 3   Temperature (C)           96453 non-null  float64
 4   Apparent Temperature (C)  96453 non-null  float64
 5   Humidity                  96453 non-null  float64
 6   Wind Speed (km/h)         96453 non-null  float64
 7   Wind Bearing (degrees)    96453 non-null  float64
 8   Visibility (km)           96453 non-null  float64
 9   Loud Cover                96453 non-null  float64
 10  Pressure (millibars)      96453 non-null  float64
 11  Daily Summary             96453 non-null  object 
dtypes: float64(8), object(4)
memory usage: 8.8+ MB


In [6]:
import statsmodels.api as sm

Y = weather_df['Temperature (C)']
X = weather_df[['Humidity', 'Wind Speed (km/h)', 'Wind Bearing (degrees)', 'Pressure (millibars)']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Temperature (C)   R-squared:                       0.421
Model:                            OLS   Adj. R-squared:                  0.421
Method:                 Least Squares   F-statistic:                 1.750e+04
Date:                Mon, 10 May 2021   Prob (F-statistic):               0.00
Time:                        15:23:07   Log-Likelihood:            -3.2821e+05
No. Observations:               96453   AIC:                         6.564e+05
Df Residuals:                   96448   BIC:                         6.565e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     37.9264      0.233    162.709      0.000      37.470      38.383
Humidity                 -32.4962      0.123   -264.288      0.000     -32.737     -32.255
Wind Speed (km/h)         -0.2014      0.003    -57.557      0.000      -0.208      -0.195
Wind Bearing (degrees)     0.0040      0.000     18.463      0.000       0.004       0.004
Pressure (millibars)      -0.0007      0.000     -3.452      0.001      -0.001      -0.000
==============================================================================
Omnibus:                     3375.432   Durbin-Watson:                   0.057
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3793.297
Skew:                          -0.455   Prob(JB):                         0.00
Kurtosis:                       3.339   Cond. No.                     1.06e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.06e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [14]:
df2=weather_df.copy()

In [15]:
df2["target"]=df2["Apparent Temperature (C)"]-df2["Temperature (C)"]
df2.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain            9.472   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain            9.356   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain            9.378   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain            8.289   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain            8.756   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                     7.389     0.890             14.120   
1                     7.228     0.860             14.265   
2                     9.378     0.890              3.928   
3                     5.944     0.830             14.104   
4                     6.978     0.830             11.045   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                 251.000           15.826       0.000              1015.130   
1                 259.000           15.826       0.000              1015.630   
2                 204.000           14.957       0.000              1015.940   
3                 269.000           15.826       0.000              1016.410   
4                 259.000           15.826       0.000              1016.510   

                       Daily Summary  target  
0  Partly cloudy throughout the day.  -2.083  
1  Partly cloudy throughout the day.  -2.128  
2  Partly cloudy throughout the day.   0.000  
3  Partly cloudy throughout the day.  -2.344  
4  Partly cloudy throughout the day.  -1.778

In [16]:
Y2 = df2['target']
X2 = df2[['Humidity', 'Wind Speed (km/h)']]

X2 = sm.add_constant(X2)

results2 = sm.OLS(Y2, X2).fit()

results2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 target   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Mon, 10 May 2021   Prob (F-statistic):               0.00
Time:                        15:39:59   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 2.4381      0.021    115.948      0.000       2.397       2.479
Humidity             -3.0292      0.024   -126.479      0.000      -3.076      -2.982
Wind Speed (km/h)    -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---

### R-squared value is decreased from 0.421 to 0.288, which means that our second model is performing worse than the first one. This tells us that only 2 explonatory variable are not enough to create a model to predict the targeted value.

---


In [18]:
df2["hum_windspeed"]=df2["Humidity"]*df2["Wind Speed (km/h)"]
df2.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain            9.472   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain            9.356   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain            9.378   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain            8.289   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain            8.756   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                     7.389     0.890             14.120   
1                     7.228     0.860             14.265   
2                     9.378     0.890              3.928   
3                     5.944     0.830             14.104   
4                     6.978     0.830             11.045   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                 251.000           15.826       0.000              1015.130   
1                 259.000           15.826       0.000              1015.630   
2                 204.000           14.957       0.000              1015.940   
3                 269.000           15.826       0.000              1016.410   
4                 259.000           15.826       0.000              1016.510   

                       Daily Summary  target  hum_windspeed  
0  Partly cloudy throughout the day.  -2.083         12.567  
1  Partly cloudy throughout the day.  -2.128         12.268  
2  Partly cloudy throughout the day.   0.000          3.496  
3  Partly cloudy throughout the day.  -2.344         11.706  
4  Partly cloudy throughout the day.  -1.778          9.167

In [19]:
Y2 = df2['target']
X2 = df2[['Humidity', 'Wind Speed (km/h)', 'hum_windspeed']]

X2 = sm.add_constant(X2)

results2 = sm.OLS(Y2, X2).fit()

results2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 target   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Mon, 10 May 2021   Prob (F-statistic):               0.00
Time:                        15:49:17   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 0.0839      0.033      2.511      0.012       0.018       0.149
Humidity              0.1775      0.043      4.133      0.000       0.093       0.262
Wind Speed (km/h)     0.0905      0.002     36.797      0.000       0.086       0.095
hum_windspeed        -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

---

### R-squared value is increased from 0.288 to 0.341, so we have a better model now. Coefficints of Humidity and Wind Speed have changed signs from negative to positive. Also the value of all coefficients, including the constant, has became smaller. 

---